README.md
## Web3.js Interface 

Interact with nodes through web3 objects provided by web3.js. On the underlying implementation, it communicates with the local node through RPC calls. web3.js can connect to any PlatON node that exposes the RPC interface.

### Usage

First, make sure the nodeJS environment is successfully installed locally. `WEB3.JS` uses the [lerna](https://github.com/lerna/lerna) management tool to optimize the workflow of the multi-package code base hosted on `git/npm`, so you should make sure the lerna package has been installed globally before installing. If not, execute the command `npm i lerna -g` for global installation. 

Then you can integrate client-sdk-js into the project through package management tools such as npm or yarn, the steps are as follows:

- npm: `npm i PlatONnetwork/client-sdk-js#0.15.1-develop`
- yarn: `yarn add PlatONnetwork/client-sdk-js#0.15.1-develop`

Create a web3 instance and set up a provider. You can refer to the following code:

```js
// in node.js
var Web3 = require('web3');

var web3 = new Web3('http://127.0.0.1:6789');
console.log(web3);
> {
    platon: ... ,
    utils: ...,
    ppos: ...,
    ...
}
```

After successful introduction, you can use the relevant API of web3.


### API Reference

#### web3.version

`web3.version`: Contains the current package version of the web3.js library.

Method:

```js
Web3.version
web3.version
```

Returns:

`String`: Current version number.

Example:

```js
web3.version;
> "0.13.1"
```

***

#### web3.modules

`web3.modules` Will return an object with the classes of all major sub modules, to be able to instantiate them manually.

Method:

```js
Web3.modules
web3.modules
```

Returns:

`Object`: A list of module constructors:

*  `Platon` - Function: The PlatON module for interacting with the PlatON network see web3.platon for more.
*  `Net` - Function: The Net module for interacting with network properties see web3.platon.net for more.
*  `Personal` - Function: The Personal module for interacting with the PlatON accounts see web3.platon.personal for more.

Example:

```js
web3.modules
> {
    Platon: Platon function(provider),
    Net: Net function(provider),
    Personal: Personal function(provider)
}
```

***

#### web3.setProvider

`web3.setProvider()` Will change the provider for it's module.

Method:

```
web3.setProvider(myProvider)
web3.platon.setProvider(myProvider)
...
```

Notes: When called on the umbrella package web3 it will also set the provider for all sub modules web3.platon, web3.shh, etc EXCEPT web3.bzz which needs a separate provider at all times.

Parameter:

`Object` - `myProvider`: a valid provider.

Returns:

`Boolean`

Example:

```js
var Web3 = require('web3');
var web3 = new Web3('http://localhost:8545');
// or
var web3 = new Web3(new Web3.providers.HttpProvider('http://localhost:8545'));

// change provider
web3.setProvider('ws://localhost:8546');
// or
web3.setProvider(new Web3.providers.WebsocketProvider('ws://localhost:8546'));
```

***

#### web3.providers

Contains the current available providers.

Method:

```
web3.providers
web3.platon.providers
...
```

Returns:

`Object`， with the following providers:

* `Object` - `HttpProvider`: The HTTP provider is deprecated, as it won’t work for subscriptions.
* `Object` - `WebsocketProvider`: The Websocket provider is the standard for usage in legacy browsers.

Example:

```js
var Web3 = require('web3');
// // use the given Provider, e.g in Mist, or instantiate a new websocket provider.
var web3 = new Web3(Web3.givenProvider || 'ws://remotenode.com:8546');
// or
var web3 = new Web3(Web3.givenProvider || new Web3.providers.WebsocketProvider('ws://remotenode.com:8546'));
```

***

#### web3.givenProvider

When using web3.js in an PlatON compatible browser, it will set with the current native provider by that browser. Will return the given provider by the (browser) environment, otherwise null.

Method:

```
web3.givenProvider
web3.platon.givenProvider
...
```

Returns:

`Object`:  The given provider set or null;

***

#### web3.currentProvider

`web3.currentProvider` Will return the current provider, otherwise null.

Method:

```
web3.currentProvider
web3.platon.currentProvider
...
```

Returns:

`Object`： The current provider set or null;

***

#### web3.BatchRequest

`web3.BatchRequest` Class to create and execute batch requests.

Method:

```
new web3.BatchRequest()
new web3.platon.BatchRequest()
```

Parameter:

none


Returns:

`Object`: With the following methods:

*  `add(request)`: To add a request object to the batch call.
*  `execute()`: Will execute the batch request.

Example:

```js
var contract = new web3.platon.Contract(abi, address);

var batch = new web3.BatchRequest();
batch.add(web3.platon.getBalance.request('lax1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqmscn5j', 'latest', callback));
batch.add(contract.methods.balance(address).call.request({from: 'lax1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqmscn5j'}, callback2));
batch.execute();
```

***

#### web3.platon.defaultAccount

`web3.platon.defaultAccount` This default address is used as the default "from" property, if no "from" property is specified in the following methods:

*  web3.platon.sendTransaction()
*  web3.platon.call()
*  new web3.platon.Contract() -> myContract.methods.myMethod().call()
*  new web3.platon.Contract() -> myContract.methods.myMethod().send()

Method:

```
web3.platon.defaultAccount
```

Property：

`String` - 20 Bytes: 20 Bytes: Any PlatON address. You should have the private key for that address in your node or keystore. (Default is undefined)

Example:

```js
web3.platon.defaultAccount;
> undefined

// set the default account
web3.platon.defaultAccount = 'atx1fyeszufxwxk62p46djncj86rd553skpph926ws';
```
> Note:
>
> **All addresses must be quoted, as above: 'lax1fyeszufxwxk62p46djncj86rd553skpptsj8v6'.**
***
#### web3.platon.defaultBlock

`web3.platon.defaultBlock` The default block is used for certain methods. 

*  web3.platon.getBalance()
*  web3.platon.getCode()
*  web3.platon.getTransactionCount()
*  web3.platon.getStorageAt()
*  web3.platon.call()
*  new web3.platon.Contract() -> myContract.methods.myMethod().call()

You can override it by passing in the defaultBlock as last parameter. 

Method:

```
web3.platon.defaultBlock
```

Property:

Default block parameters can be one of the following:

*  `Number`: A block number
*  `"genesis"` - `String`: The genesis block
*  `"latest"` - `String`: The latest block (current head of the blockchain)
*  `"pending"` - `String`: The currently mined block (including pending transactions)

Default is "latest"

Example:

```js
web3.platon.defaultBlock;
> "latest"

// set the default block
web3.platon.defaultBlock = 231;
```

***

#### web3.platon.getProtocolVersion

Returns the PlatON protocol version of the node.

Method:

```
web3.platon.getProtocolVersion([callback])
```

Returns:

`Promise` returns `String`: the protocol version.

Example:

```js
web3.platon.getProtocolVersion().then(console.log);
> "63"
```

***

#### web3.platon.isSyncing

`web3.platon.isSyncing()` Checks if the node is currently syncing and returns either a syncing object, or false.

Method:

```
web3.platon.isSyncing([callback])
```

Returns:

`Promise` returns `Object|Boolean` - A sync object when the node is currently syncing or false:

* `startingBlock` - `Number`: The block number where the sync started.
* `currentBlock` - `Number`: The block number where at which block the node currently synced to already.
* `highestBlock` - `Number`: The estimated block number to sync to.
* `knownStates` - `Number`: The estimated states to download.
* `pulledStates` - `Number`: The already downloaded states.

Example:

```js
web3.platon.isSyncing().then(console.log);
> {
    startingBlock: 100,
    currentBlock: 312,
    highestBlock: 512,
    knownStates: 234566,
    pulledStates: 123455
}
```

***

#### web3.platon.getGasPrice

`web3.platon.getGasPrice()` Returns the current gas price oracle. The gas price is determined by the last few blocks median gas price.

Method:

```
web3.platon.getGasPrice([callback])
```

Returns:

`Promise` returns `String` - Number string of the current gas price in von.

Example:

```js
web3.platon.getGasPrice().then(console.log);
> "20000000000"
```

***

#### web3.platon.getAccounts

`web3.platon.getAccounts()` Returns a list of accounts the node controls.

Method:

```
web3.platon.getAccounts([callback])
```

Returns:

`Promise` returns `Array` - An array of addresses controlled by node.

Example:

```js
web3.platon.getAccounts().then(console.log);
> ["atx1fyeszufxwxk62p46djncj86rd553skpph926ws", "atx1kg7y7wfwzqsyxppyxcdvhkkkwlf64cclmnav7p"]
```

***

#### web3.platon.getBlockNumber

`web3.platon.getBlockNumber()` Returns the current block number.

Method:

```
web3.platon.getBlockNumber([callback])
```

Returns:

`Promise` returns `Number` - The number of the most recent block.

Example:

```js
web3.platon.getBlockNumber().then(console.log);
> 2744
```

***

#### web3.platon.getBalance

`web3.platon.getBalance()` Get the balance of an address at a given block.

Method:

```
web3.platon.getBalance(address [, defaultBlock] [, callback])
```

Parameter:

*  `address`：String - The address to get the balance of.
*  `defaultBlock`：Number|String -  (optional) If you pass this parameter it will not use the default block set with web3.platon.defaultBlock. Pre-defined block numbers as "latest", "earliest", "pending", and "genesis" can also be used.
*  `callback`：Function - (optional) Optional callback, returns an error object as first parameter and the result as second.

Returns:

`Promise` returns `String` - The current balance for the given address in von.

Example:

```js
web3.platon.getBalance("atx1fyeszufxwxk62p46djncj86rd553skpph926ws")
.then(console.log);
> "1000000000000"
```

***

#### web3.platon.getStorageAt

`web3.platon.getStorageAt()` Get the storage at a specific position of an address.

Method:

```
web3.platon.getStorageAt(address, position [, defaultBlock] [, callback])
```

Parameter:

*  `address` - `String`: The address to get the storage from.
*  `position` - `Number`: The index position of the storage.
*  `defaultBlock` -`Number|String` (optional) If you pass this parameter it will not use the default block set with web3.platon.defaultBlock. Pre-defined block numbers as "latest", "earliest", "pending", and "genesis" can also be used.
*  `callback` -`Function`: (optional) Optional callback, returns an error object as first parameter and the result as second.

Returns:

`Promise` returns `String` - The value in storage at the given position.

Example:

```js
web3.platon.getStorageAt("atx1fyeszufxwxk62p46djncj86rd553skpph926ws", 0)
.then(console.log);
> "0x033456732123ffff2342342dd12342434324234234fd234fd23fd4f23d4234"
```

***

#### web3.platon.getCode

`web3.platon.getCode()` Get the code at a specific address.

Method:

```
web3.platon.getCode(address [, defaultBlock] [, callback])
```

Parameter:

*  `address` - `String`: The address to get the code from.
*  `defaultBlock` - `Number|String`: (optional) If you pass this parameter it will not use the default block set with web3.platon.defaultBlock. Pre-defined block numbers as "latest", "earliest", "pending", and "genesis" can also be used.
*  `callback` - `Function`: (optional) Optional callback, returns an error object as first parameter and the result as second.

Returns:

`Promise` returns `String` - The data at given address.

Example:

```js
web3.platon.getCode("atx1fyeszufxwxk62p46djncj86rd553skpph926ws")
.then(console.log);
> "0x600160008035811a818181146012578301005b601b6001356025565b8060005260206000f25b600060078202905091905056"
```

***

#### web3.platon.getBlock

`web3.platon.getBlock()` Returns a block matching the block number or block hash.

Method:

```
web3.platon.getBlock(blockHashOrBlockNumber [, returnTransactionObjects] [, callback])
```

Parameter:

*  `blockHashOrBlockNumber` - `String|Number`: The block number or block hash. Or the string "genesis", "latest", "earliest", or "pending" as in the default block parameter.
*  `returnTransactionObjects` - `Boolean`: (optional, default false) If specified true, the returned block will contain all transactions as objects. By default it is false so, there is no need to explictly specify false. And, if false it will only contains the transaction hashes.
*  `callback` - `Function`: (optional) Optional callback, returns an error object as first parameter and the result as second.

Returns:

Promise returns Object - The block object:

* `number` - `Number`: The block number. null when it's pending block.
* `hash` 32 Bytes - `String`: Hash of the block. null when it's pending block.
* `parentHash` 32 Bytes - `String`: Hash of the parent block.
* `nonce` 8 Bytes - `String`: Hash of the generated proof-of-work. null when it's pending block.
* `sha3Uncles` 32 Bytes - `String`: SHA3 of the uncles data in the block.
* `logsBloom` 256 Bytes - `String`: The bloom filter for the logs of the block. null when it's pending block.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- EXAMPLE

## [1.0.0]

### Added

- I've added feature XY (#1000)

### Changed

- I've cleaned up XY (#1000)

### Deprecated

- I've deprecated XY (#1000)

### Removed

- I've removed XY (#1000)

### Fixed

- I've fixed XY (#1000)

### Security

- I've improved the security in XY (#1000)

-->

## [1.2.0]

Released with 1.0.0-beta.37 code base.

## [1.2.1]

### Fixed

- ``websocket`` dependency fixed (#2971, #2976)
- ``requestOptions`` added to ``WebsocketProvider`` (#2979)
- Node >= v8.0.0 support (#2938)

## [1.2.2]

### Added

- localStorage support detection added (#3031)
- getNetworkType method extended with Görli testnet (#3095)
- supportsSubscriptions method added to providers (#3116)
- Add `eth.getChainId` method (#3113)
- Minified file added to web3 package (#3131)
- The transaction confirmation workflow can now be configured (#3130)
- Additional parameters for accounts.signTransaction added [(docs)](https://web3js.readthedocs.io/en/v1.2.2/web3-eth-accounts.html#signtransaction) (#3141)
- Emit `connected` event on subscription creation (#3028)
- TypeScript type definitions added for all modules (#3132)
- Bloom filters added to web3.utils (#3137)

### Fixed

- Fix allow `0` as a valid `fromBlock` or `toBlock` filter param (#1100)
- Fix randomHex returning inconsistent string lengths (#1490)
- Fix make isBN minification safe (#1777)
- Fix incorrect references to BigNumber in utils.fromWei and utils.toWei error messages (#2468)
- Fix error incorrectly thrown when receipt.status is `null` (#2183)
- Fix incorrectly populating chainId param with `net_version` when signing txs (#2378)
- regeneratorRuntime error fixed (#3058)
- Fix accessing event.name where event is undefined (#3014)
- fixed Web3Utils toHex() for Buffer input (#3021)
- Fix bubbling up tx signing errors (#2063, #3105)
- HttpProvider: CORS issue with Firefox and Safari (#2978)
- Ensure the immutability of the `tx` object passed to function `signTransaction` (#2190)
- Gas check fixed (#2381)
- Signing issues #1998, #2033, and #1074 fixed (#3125)
- Fix hexToNumber and hexToNumberString prefix validation (#3086)
- The receipt will now returned on a EVM error (this got removed on beta.18) (#3129)
- Fixes transaction confirmations with the HttpProvider (#3140)

## [1.2.3]

### Fixed

- Fix perfect gas usage causes tx to error (#3175)
- Fix regenerator runtime error in web3.min.js (#3155)
- Fix TS types for eth.subscribe syncing, newBlockHeaders, pendingTransactions (#3159)
- Improve web3-eth-abi decodeParameters error message (#3134)

## [1.2.4]

### Fixed

- Fix npm installation error for scrypt-shim and websocket (#3210)

## [Unreleased]

## [1.2.5]

### Added

- ``eth_requestAccounts`` as ``requestAccounts`` added to web3-eth package (#3219)
- ``sha3Raw`` and ``soliditySha3Raw`` added to web3-utils package (#3226)
- ``eth_getProof`` as ``getProof`` added to web3-eth package (#3220)
- ``BN`` and ``BigNumber`` objects are now supported by the ``abi.encodeParameter(s)`` method (#3238)
- ``getPendingTransactions`` added to web3-eth package (#3239)
- Revert instruction handling added which can get activated with the ``handleRevert`` module property (#3248)
- The ``receipt`` does now exist as property on the error object for transaction related errors (#3259)
- ``internalType`` added to ``AbiInput`` TS interface in ``web3-utils`` (#3279)

### Changed

- ``eth-lib`` dependency updated (0.2.7 => ^0.2.8) (#3242)

### Fixed

- Fix user supplied callback not fired in eth.accounts.signTransaction (#3283)
- Fix minified bundle (#3256)
- ``defaultBlock`` property handling fixed (#3247)
- ``clearSubscriptions`` does no longer throw an error if no running subscriptions do exist (#3246)
- callback type definition for ``Accounts.signTransaction`` fixed (#3280)
- fix: export bloom functions on the index.js

### Fixed

- new params: PPOS function: 1004 delegate() typ:3
- new return: PPOS Api: include(1004(delegate)/1005(withdrewDelegation)/1104(getDelegateInfo)/1106(getDelegationLockInfo))
- fix bug: If the return value 'logs Data' has data(type:1005/1006/5000), but has not been processed before

lib/utils/utils.js
/*
    This file is part of web3.js.

    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file utils.js
 * @author Marek Kotewicz <marek@ethdev.com>
 * @date 2015
 */

/**
 * Utils
 *
 * @module utils
 */

/**
 * Utility functions
 *
 * @class [utils] utils
 * @constructor
 */


var BigNumber = require('bignumber.js');
var sha3 = require('./sha3.js');
var utf8 = require('utf8');
var segwit_addr = require('./segwit_addr.js');

var unitMap = {
    'nolat':      '0',
    'von':          '1',
    'kvon':         '1000',
    'Kvon':         '1000',
    'babbage':      '1000',
    'femtolat':   '1000',
    'mvon':         '1000000',
    'Mvon':         '1000000',
    'lovelace':     '1000000',
    'picolat':    '1000000',
    'gvon':         '1000000000',
    'Gvon':         '1000000000',
    'shannon':      '1000000000',
    'nanolat':    '1000000000',
    'nano':         '1000000000',
    'szabo':        '1000000000000',
    'microlat':   '1000000000000',
    'micro':        '1000000000000',
    'finney':       '1000000000000000',
    'millilat':   '1000000000000000',
    'milli':        '1000000000000000',
    'lat':        '1000000000000000000',
    'klat':       '1000000000000000000000',
    'grand':        '1000000000000000000000',
    'mlat':       '1000000000000000000000000',
    'glat':       '1000000000000000000000000000',
    'tlat':       '1000000000000000000000000000000'
};

/**
 * Should be called to pad string to expected length
 *
 * @method padLeft
 * @param {String} string to be padded
 * @param {Number} characters that result string should have
 * @param {String} sign, by default 0
 * @returns {String} right aligned string
 */
var padLeft = function (string, chars, sign) {
    return new Array(chars - string.length + 1).join(sign ? sign : "0") + string;
};

/**
 * Should be called to pad string to expected length
 *
 * @method padRight
 * @param {String} string to be padded
 * @param {Number} characters that result string should have
 * @param {String} sign, by default 0
 * @returns {String} right aligned string
 */
var padRight = function (string, chars, sign) {
    return string + (new Array(chars - string.length + 1).join(sign ? sign : "0"));
};

/**
 * Should be called to get utf8 from it's hex representation
 *
 * @method toUtf8
 * @param {String} string in hex
 * @returns {String} ascii string representation of hex value
 */
var toUtf8 = function(hex) {
// Find termination
    var str = "";
    var i = 0, l = hex.length;
    if (hex.substring(0, 2) === '0x') {
        i = 2;
    }
    for (; i < l; i+=2) {
        var code = parseInt(hex.substr(i, 2), 16);
        if (code === 0)
            break;
        str += String.fromCharCode(code);
    }

    return utf8.decode(str);
};

/**
 * Should be called to get ascii from it's hex representation
 *
 * @method toAscii
 * @param {String} string in hex
 * @returns {String} ascii string representation of hex value
 */
var toAscii = function(hex) {
// Find termination
    var str = "";
    var i = 0, l = hex.length;
    if (hex.substring(0, 2) === '0x') {
        i = 2;
    }
    for (; i < l; i+=2) {
        var code = parseInt(hex.substr(i, 2), 16);
        str += String.fromCharCode(code);
    }

    return str;
};

/**
 * Should be called to get hex representation (prefixed by 0x) of utf8 string
 *
 * @method fromUtf8
 * @param {String} string
 * @param {Boolean} allowZero to convert code point zero to 00 instead of end of string
 * @returns {String} hex representation of input string
 */
var fromUtf8 = function(str, allowZero) {
    str = utf8.encode(str);
    var hex = "";
    for(var i = 0; i < str.length; i++) {
        var code = str.charCodeAt(i);
        if (code === 0) {
            if (allowZero) {
                hex += '00';
            } else {
                break;
            }
        } else {
            var n = code.toString(16);
            hex += n.length < 2 ? '0' + n : n;
        }
    }

    return "0x" + hex;
};

/**
 * Should be called to get hex representation (prefixed by 0x) of ascii string
 *
 * @method fromAscii
 * @param {String} string
 * @param {Number} optional padding
 * @returns {String} hex representation of input string
 */
var fromAscii = function(str) {
    var hex = "";
    for(var i = 0; i < str.length; i++) {
        var code = str.charCodeAt(i);
        var n = code.toString(16);
        hex += n.length < 2 ? '0' + n : n;
    }

    return "0x" + hex;
};

/**
 * Should be used to create full function/event name from json abi
 *
 * @method transformToFullName
 * @param {Object} json-abi
 * @return {String} full fnction/event name
 */
var transformToFullName = function (json) {
    if (json.name.indexOf('(') !== -1) {
        return json.name;
    }

    var typeName = json.inputs.map(function(i){return i.type; }).join();
    return json.name + '(' + typeName + ')';
};

/**
 * Should be called to get display name of contract function
 *
 * @method extractDisplayName
 * @param {String} name of function/event
 * @returns {String} display name for function/event eg. multiply(uint256) -> multiply
 */
var extractDisplayName = function (name) {
    var stBracket = name.indexOf('(');
    var endBracket = name.indexOf(')');
    return (stBracket !== -1 && endBracket !== -1) ? name.substr(0, stBracket) : name;
};

/**
 * Should be called to get type name of contract function
 *
 * @method extractTypeName
 * @param {String} name of function/event
 * @returns {String} type name for function/event eg. multiply(uint256) -> uint256
 */
var extractTypeName = function (name) {
    var stBracket = name.indexOf('(');
    var endBracket = name.indexOf(')');
    return (stBracket !== -1 && endBracket !== -1) ? name.substr(stBracket + 1, endBracket - stBracket - 1).replace(' ', '') : "";
};

/**
 * Converts value to it's decimal representation in string
 *
 * @method toDecimal
 * @param {String|Number|BigNumber}
 * @return {String}
 */
var toDecimal = function (value) {
    return toBigNumber(value).toNumber();
};

/**
 * Converts value to it's hex representation
 *
 * @method fromDecimal
 * @param {String|Number|BigNumber}
 * @return {String}
 */
var fromDecimal = function (value) {
    var number = toBigNumber(value);
    var result = number.toString(16);

    return number.lessThan(0) ? '-0x' + result.substr(1) : '0x' + result;
};

/**
 * Auto converts any given value into it's hex representation.
 *
 * And even stringifys objects before.
 *
 * @method toHex
 * @param {String|Number|BigNumber|Object}
 * @return {String}
 */
var toHex = function (val) {
    /*jshint maxcomplexity: 8 */

    if (isBoolean(val))
        return fromDecimal(+val);

    if (isBigNumber(val))
        return fromDecimal(val);

    if (typeof val === 'object')
        return fromUtf8(JSON.stringify(val));

    // if its a negative number, pass it through fromDecimal
    if (isString(val)) {
        if (val.indexOf('-0x') === 0)
            return fromDecimal(val);
        else if(val.indexOf('0x') === 0)
            return val;
        else if (!isFinite(val))
            return fromUtf8(val,1);
    }

    return fromDecimal(val);
};

/**
 * Returns value of unit in von
 *
 * @method getValueOfUnit
 * @param {String} unit the unit to convert to, default lat
 * @returns {BigNumber} value of the unit (in von)
 * @throws error if the unit is not correct:w
 */
var getValueOfUnit = function (unit) {
    unit = unit ? unit.toLowerCase() : 'lat';
    var unitValue = unitMap[unit];
    if (unitValue === undefined) {
        throw new Error('This unit doesn\'t exists, please use the one of the following units' + JSON.stringify(unitMap, null, 2));
    }
    return new BigNumber(unitValue, 10);
};

/**
 * Takes a number of von and converts it to any other lat unit.
 *
 * Possible units are:
 *   SI Short   SI Full        Effigy       Other
 * - kvon       femtolat     babbage
 * - mvon       picolat      lovelace
 * - gvon       nanolat      shannon      nano
 * - --         microlat     szabo        micro
 * - --         millilat     finney       milli
 * - lat      --             --
 * - klat                    --           grand
 * - mlat
 * - glat
 * - tlat
 *
 * @method fromVon
 * @param {Number|String} number can be a number, number string or a HEX of a decimal
 * @param {String} unit the unit to convert to, default lat
 * @return {String|Object} When given a BigNumber object it returns one as well, otherwise a number
*/
var fromVon = function(number, unit) {
    var returnValue = toBigNumber(number).dividedBy(getValueOfUnit(unit));

    return isBigNumber(number) ? returnValue : returnValue.toString(10);
};

/**
 * Takes a number of a unit and converts it to von.
 *
 * Possible units are:
 *   SI Short   SI Full        Effigy       Other
 * - kvon       femtolat     babbage
 * - mvon       picolat      lovelace
 * - gvon       nanolat      shannon      nano
 * - --         microlat     szabo        micro
 * - --         millilat     finney       milli
 * - lat      --             --
 * - klat                    --           grand
 * - mlat
 * - glat
 * - tlat
 *
 * @method toVon
 * @param {Number|String|BigNumber} number can be a number, number string or a HEX of a decimal
 * @param {String} unit the unit to convert from, default lat
 * @return {String|Object} When given a BigNumber object it returns one as well, otherwise a number
*/
var toVon = function(number, unit) {
    var returnValue = toBigNumber(number).times(getValueOfUnit(unit));

    return isBigNumber(number) ? returnValue : returnValue.toString(10);
};

/**
 * Takes an input and transforms it into an bignumber
 *
 * @method toBigNumber
 * @param {Number|String|BigNumber} a number, string, HEX string or BigNumber
 * @return {BigNumber} BigNumber
*/
var toBigNumber = function(number) {
    /*jshint maxcomplexity:5 */
    number = number || 0;
    if (isBigNumber(number))
        return number;

    if (isString(number) && (number.indexOf('0x') === 0 || number.indexOf('-0x') === 0)) {
        return new BigNumber(number.replace('0x',''), 16);
    }

    return new BigNumber(number.toString(10), 10);
};

/**
 * Takes and input transforms it into bignumber and if it is negative value, into two's complement
 *
 * @method toTwosComplement
 * @param {Number|String|BigNumber}
 * @return {BigNumber}
 */
var toTwosComplement = function (number) {
    var bigNumber = toBigNumber(number).round();
    if (bigNumber.lessThan(0)) {
        return new BigNumber("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff", 16).plus(bigNumber).plus(1);
    }
    return bigNumber;
};

/**
 * Checks if the given string is strictly an address
 *
 * @method isStrictAddress
 * @param {String} address the given HEX adress
 * @return {Boolean}
*/
var isStrictAddress = function (address) {
    return /^0x[0-9a-f]{40}$/i.test(address);
};

/**
 * Checks if the given string is an address
 *
 * @method isAddress
 * @param {String} address the given HEX adress
 * @return {Boolean}
*/
var isAddress = function (address) {
    if (!/^(0x)?[0-9a-f]{40}$/i.test(address)) {
        // check if it has the basic requirements of an address
        return false;
    } else if (/^(0x)?[0-9a-f]{40}$/.test(address) || /^(0x)?[0-9A-F]{40}$/.test(address)) {
        // If it's all small caps or all all caps, return true
        return true;
    } else {
        // Otherwise check each case
        return isChecksumAddress(address);
    }
};

/**
 * Checks if the given string is an bech32 address
 *
 * @method isBech32Address
 * @param {String} address the given bech32 adress
 * @return {Boolean}
*/
var isBech32Address = function (address) {
    if(address.length != 42)
    {
        return false;
    }
    var hrp = address.substr(0,3);
    var ret = segwit_addr.decode(hrp, address);
    if (ret === null) {
        return false;
    }
    return true;
};

/**
 * Transforms given string to bech32 addres
 *
 * @method toBech32Address
 * @param {String} hrp
 * @param {String} address
 * @return {String} formatted bech32 address
 */
var toBech32Address = function (hrp, address) {
    if (isStrictAddress(address) || isChecksumAddress(address)) {
        return segwit_addr.EncodeAddress(hrp, address);
    }

    return ''
};

/**
 * Resolve the bech32 address
 *
 * @method decodeBech32Address
 * @param {String} bech32Address
 * @return {String} formatted address
 */
var decodeBech32Address = function (bech32Address) {
    if (isBech32Address(bech32Address)) {
        var hrp = bech32Address.substr(0,3);
        address = segwit_addr.DecodeAddress(hrp, bech32Address);
        if (address !== null) {
            return "0x" + address
        }
    }

    return ''
};


/**
 * Checks if the given string is a checksummed address
 *
 * @method isChecksumAddress
 * @param {String} address the given HEX adress
 * @return {Boolean}
*/
var isChecksumAddress = function (address) {
    // Check each case
    address = address.replace('0x','');
    var addressHash = sha3(address.toLowerCase());

    for (var i = 0; i < 40; i++ ) {
        // the nth letter should be uppercase if the nth digit of casemap is 1
        if ((parseInt(addressHash[i], 16) > 7 && address[i].toUpperCase() !== address[i]) || (parseInt(addressHash[i], 16) <= 7 && address[i].toLowerCase() !== address[i])) {
            return false;
        }
    }
    return true;
};



/**
 * Makes a checksum address
 *
 * @method toChecksumAddress
 * @param {String} address the given HEX adress
 * @return {String}
*/
var toChecksumAddress = function (address) {
    if (typeof address === 'undefined') return '';

    address = address.toLowerCase().replace('0x','');
    var addressHash = sha3(address);
    var checksumAddress = '0x';

    for (var i = 0; i < address.length; i++ ) {
        // If ith character is 9 to f then make it uppercase
        if (parseInt(addressHash[i], 16) > 7) {
          checksumAddress += address[i].toUpperCase();
        } else {
            checksumAddress += address[i];
        }
    }
    return checksumAddress;
};

/**
 * Transforms given string to valid 20 bytes-length addres with 0x prefix
 *
 * @method toAddress
 * @param {String} address
 * @return {String} formatted address
 */
var toAddress = function (address) {
    if (isStrictAddress(address)) {
        return address;
    }

    if (/^[0-9a-f]{40}$/.test(address)) {
        return '0x' + address;
    }

    return '0x' + padLeft(toHex(address).substr(2), 40);
};

/**
 * Returns true if object is BigNumber, otherwise false
 *
 * @method isBigNumber
 * @param {Object}
 * @return {Boolean}
 */
var isBigNumber = function (object) {
    return object instanceof BigNumber ||
        (object && object.constructor && object.constructor.name === 'BigNumber');
};

/**
 * Returns true if object is string, otherwise false
 *
 * @method isString
 * @param {Object}
 * @return {Boolean}
 */
var isString = function (object) {
    return typeof object === 'string' ||
        (object && object.constructor && object.constructor.name === 'String');
};

/**
 * Returns true if object is function, otherwise false
 *
 * @method isFunction
 * @param {Object}
 * @return {Boolean}
 */
var isFunction = function (object) {
    return typeof object === 'function';
};

/**
 * Returns true if object is Objet, otherwise false
 *
 * @method isObject
 * @param {Object}
 * @return {Boolean}
 */
var isObject = function (object) {
    return object !== null && !(Array.isArray(object)) && typeof object === 'object';
};

/**
 * Returns true if object is boolean, otherwise false
 *
 * @method isBoolean
 * @param {Object}
 * @return {Boolean}
 */
var isBoolean = function (object) {
    return typeof object === 'boolean';
};

/**
 * Returns true if object is array, otherwise false
 *
 * @method isArray
 * @param {Object}
 * @return {Boolean}
 */
var isArray = function (object) {
    return Array.isArray(object);
};

/**
 * Returns true if given string is valid json object
 *
 * @method isJson
 * @param {String}
 * @return {Boolean}
 */
var isJson = function (str) {
    try {
        return !!JSON.parse(str);
    } catch (e) {
        return false;
    }
};

/**
 * Returns true if given string is a valid Ethereum block header bloom.
 *
 * @method isBloom
 * @param {String} hex encoded bloom filter
 * @return {Boolean}
 */
var isBloom = function (bloom) {
    if (!/^(0x)?[0-9a-f]{512}$/i.test(bloom)) {
        return false;
    } else if (/^(0x)?[0-9a-f]{512}$/.test(bloom) || /^(0x)?[0-9A-F]{512}$/.test(bloom)) {
        return true;
    }
    return false;
};

/**
 * Returns true if given string is a valid log topic.
 *
 * @method isTopic
 * @param {String} hex encoded topic
 * @return {Boolean}
 */
var isTopic = function (topic) {
    if (!/^(0x)?[0-9a-f]{64}$/i.test(topic)) {
        return false;
    } else if (/^(0x)?[0-9a-f]{64}$/.test(topic) || /^(0x)?[0-9A-F]{64}$/.test(topic)) {
        return true;
    }
    return false;
};

module.exports = {
    padLeft: padLeft,
    padRight: padRight,
    toHex: toHex,
    toDecimal: toDecimal,
    fromDecimal: fromDecimal,
    toUtf8: toUtf8,
    toAscii: toAscii,
    fromUtf8: fromUtf8,
    fromAscii: fromAscii,
    transformToFullName: transformToFullName,
    extractDisplayName: extractDisplayName,
    extractTypeName: extractTypeName,
    toVon: toVon,
    fromVon: fromVon,
    toBigNumber: toBigNumber,
    toTwosComplement: toTwosComplement,
    toAddress: toAddress,
    isBigNumber: isBigNumber,
    isStrictAddress: isStrictAddress,
    isAddress: isAddress,
    isChecksumAddress: isChecksumAddress,
    toChecksumAddress: toChecksumAddress,
    isBech32Address:isBech32Address,
    toBech32Address:toBech32Address,
    decodeBech32Address:decodeBech32Address,
    isFunction: isFunction,
    isString: isString,
    isObject: isObject,
    isBoolean: isBoolean,
    isArray: isArray,
    isJson: isJson,
    isBloom: isBloom,
    isTopic: isTopic,
};
